### ISSUE TITLE
Selecting a area of an NdOverlay of Contours and Image, why doesn't this work?

### ISSUE THREAD

**Issue by basnijholt** [ioam/holoviews#76]
```python
%%opts Layout [fig_inches=(10,10) vspace=0]
angles = {'key_dimensions': [hv.Dimension(r'$\phi$', unit=r'$\pi$ rad'), hv.Dimension(r'$\theta$', unit=r'$\pi$ rad')]}

a = lambda: hv.Image(np.random.randn(10,10), bounds=(0, 0, 10, 10))
b = lambda: hv.Contours([zip(range(11), np.sort(np.random.rand(11)*10))]) * hv.Contours([zip(range(11), np.sort(np.random.rand(11)*10))])
c = {(1,1): a(),
     (1,2): a(),
     (2,1): a(),
     (2,2): a()}

d = {(1,1): b(),
     (1,2): b(),
     (2,1): b(),
     (2,2): b()}
(hv.HoloMap(c, **angles) * hv.HoloMap(d, **angles) * hv.Bounds((1,1,5,5))
+ (hv.HoloMap(c, **angles) * hv.HoloMap(d, **angles)).select(x=(1, 5), y=(1, 5)))
```
![screen shot 2015-05-15 at 12 45 22](https://cloud.githubusercontent.com/assets/6897215/7651401/4d7bfdfc-fb00-11e4-9686-bb962ead2715.png)

But this doesn't work (why?):
```python
%%opts Layout [fig_inches=(10,10) vspace=0]
a = lambda: hv.Image(np.random.randn(10,10), bounds=(0, 0, 10, 10))
b = lambda: hv.NdOverlay({1: hv.Contours([zip(range(11), np.sort(np.random.rand(11)*10))]), 2: hv.Contours([zip(range(11), np.sort(np.random.rand(11)*10))])})
c = {(1,1): a(),
     (1,2): a(),
     (2,1): a(),
     (2,2): a()}

d = {(1,1): b(),
     (1,2): b(),
     (2,1): b(),
     (2,2): b()}
(hv.HoloMap(c, **angles) * hv.HoloMap(d, **angles) * hv.Bounds((1,1,5,5))
+ (hv.HoloMap(c, **angles) * hv.HoloMap(d, **angles)).select(x=(1, 5), y=(1, 5)))
```

**Comment by basnijholt** [comment]
With an ``Overlay`` instead of a ``NdOverlay`` it does work btw. 

But it selects a incorrect area:
![screen shot 2015-05-15 at 12 50 24](https://cloud.githubusercontent.com/assets/6897215/7651470/1375428e-fb01-11e4-902b-54a33e16d926.png)

**Comment by philippjfr** [comment]
This is a fairly major limitation of select that I've known about but haven't yet had time to address. Basically select is a simply wrapper around ``__getitem__``, which queries the deep_dimensions and then constructs the appropriate deep slices. This was okay as long as our data structures were homogenous, i.e. followed the same nesting structure and shared dimensions. However since introducing Layout and Overlay, you can now have very different nesting structures. The correct solution is to make calls to select recursive. If any of the supplied dimension indices apply at that level then apply apply the selection with ``__getitem__``, otherwise pass the remaining slices to all Elements as long as they contain the selected dimension. Shouldn't be too difficult to implement and it's a fairly important feature. I'll have a shot at it this weekend.

**Comment by basnijholt** [comment]
This is probably related:
```python
%%opts Contours [aspect='square']
x = np.linspace(-50,50)
y = x**2
a = hv.Contours([zip(x, y)])
b = a.select(x=(-40, 40), y=(0, 500))
b
```
![download 2](https://cloud.githubusercontent.com/assets/6897215/7661513/8adae746-fb53-11e4-9579-ecb3c1fe0a7f.png)


then
```python 
%%opts Contours [aspect='square']
b.select(x=(-10,10))
```
![download 3](https://cloud.githubusercontent.com/assets/6897215/7661519/9367e06c-fb53-11e4-8f17-5e0dcedf37ab.png)

here it forgets the original select of ``y=(0, 500)``.

**Comment by philippjfr** [comment]
This is now fixed in commit 65c9062 with new unit tests added in commit 960343d. Select now works recursively as outlined in my previous comments. This should ensure that the correct slices are always applied to the correct objects. Please do test it yourself and reopen the issue if there are any problems.

**Comment by basnijholt** [comment]
I don't understand what you mean by recursively selecting.

In case you mean that I first select the area of the ``Image`` and the separately ``Contours`` and then make an overlay, this is what happens (the left picture is what it should look like, and the left one is the selection):

```python
(hv.HoloMap(orb_zoom_lines['gaps_bounds'], **angles)[0, 0] 
 + hv.HoloMap(orb['gaps'], **angles)[0, 0].select(**{'B':(1.2, 1.95), 'Chemical potential':(8, 30)}) 
 * hv.HoloMap(orb['topo_bounds'], **angles)[0, 0].select(**{'B':(1.2, 1.95), 'Chemical potential':(8, 30)}))
```
![download 11](https://cloud.githubusercontent.com/assets/6897215/7678102/abff9070-fd51-11e4-8afc-f9307f132e36.png)


So the selection of the ``Contours`` is correct and the ``Image`` is not.

**Comment by basnijholt** [comment]
* I can't reopen the issue btw.

**Comment by philippjfr** [comment]
No recursive selection is an internal detail of how .select works it's not something you should have to worry about. The example should now work, as posted above. I'll double check it's working now, but your example looks worrying as that should be working too.

**Comment by philippjfr** [comment]
Hmm, I just tested it with this simpler example:

```python
hmap = hv.HoloMap({i: (hv.Image(np.random.rand(10, 10)) * hv.Path([np.random.rand(5,2)]))
            for i in range(3)})
hmap.select(x=(0, 0.5), y=(0, 0.5))
```

and that seemed to work fine. Could you come up with a minimal example that seems broken to you?

**Comment by basnijholt** [comment]
I'm not sure if this is it, but at least this doesn't work:
```python 
import numpy as np
import holoviews as hv
%load_ext holoviews.ipython

x,y = np.mgrid[-50:51, -50:51] * 0.1
z = np.sin(x**2 + y**2)

def circle(radius, x=0, y=0):
    angles = np.linspace(0, 2*np.pi, 100)
    return np.array( list(zip(x+radius*np.sin(angles), y+radius*np.cos(angles))))
```

```python
%%opts Image (origin='lower')
(hv.Image(z) * hv.Contours([circle(0.5)])).select(x=(0,0.5), y=(0,0.5))
```
this gives:

![download 12](https://cloud.githubusercontent.com/assets/6897215/7682516/e6e91b58-fd78-11e4-8a99-7030ed65767b.png)

```python
(hv.Image(z) * hv.Contours([circle(0.5)])).select(x=(0,0.5), y=(0,0.5))
```
this gives:
![download 13](https://cloud.githubusercontent.com/assets/6897215/7682521/ee92ce76-fd78-11e4-986d-d3e0983fdc00.png)

**Comment by basnijholt** [comment]
This will produce something incorrect:
```python
%%opts Image (origin='lower')
data = no_orb['gaps'][0,0].data
im = hv.Image(data, bounds=(0,0,2,35))
im*hv.HLine(8)*hv.VLine(1.2) + im.select(x=(1.2, 1.95), y=(8, 30)) 
```

![download 14](https://cloud.githubusercontent.com/assets/6897215/7682818/10dad9d6-fd7b-11e4-9df8-9be4370db73d.png)


I use my own data but use a raw ``np.array`` instead of randomly generated data that doesn't show it so well.

**Comment by basnijholt** [comment]
```python
%%opts Image (origin='lower')
data = np.loadtxt('testdata.dat')
hv.Image(data, bounds=(0,0,2,35)) * hv.Bounds((1.2, 8, 1.95 , 30)) + hv.Image(data, bounds=(0,0,2,30)).select(x=(1.2, 1.95), y=(8,30))
```

![download 15](https://cloud.githubusercontent.com/assets/6897215/7690689/1bd1d7f0-fdb6-11e4-98f7-98e8e0460856.png)


download the file here:

http://nijholt.biz/topo/testdata.dat

**Comment by jlstevens** [comment]
This was indeed confusing but we've figured it out. Here is what I think you ought to be doing:

```python
data = np.loadtxt('testdata.dat')
loaded = np.loadtxt('testdata.dat')
data = np.flipud(loaded)
(hv.Image(data, bounds=(0,0,2,35)) * hv.Bounds((1.2, 8, 1.95 , 30)) 
+ hv.Image(data, bounds=(0,0,2,30)).select(x=(1.2, 1.95), y=(8,30)))
```

**Note**: I changed the two bounds to match exactly - the bounds passed to ``Image`` are about the continuous space in which the data actually lives and so must match whether you then select or not!

Anyway, the key change is to get rid of the ``(origin='lower')`` option and flip the actual data with ``np.flipud``. We realize that the origin option is very confusing because it only flips the pixel locations drawn in the underlying ``imshow`` call and not the actual data!

This means that what was happening is what you were seeing did not match what you were actually slicing - the black and white pixels in A of your figure are flipped top-to-bottom. If you imagine flipping A back again (i.e the way it is actually being sliced) and then compensate for the slightly mismatched bounds, then B makes sense again!

So in conclusion, flipping the actual input data is the right thing to do here and the origin option is very  confusing as it leads to a disconnect between what you see and your actual data. I think this means we should deprecate the origin style option entirely as its semantics are unclear as it lets you fix data that isn't correct in your elements by inverting the display to make it *look* correct. Note that the ``invert_yaxis`` plot option still exists and is still valid as when it flips the displayed data it also flips the y tick labels to match.

Hopefully this explanation makes some sense to you but I think I have convinced myself that the origin option will only generate wrong results and confusion!

**Comment by jlstevens** [comment]
Removed the 'origin' style keyword for rasters in e0c0a8c as it will only cause confusion for the reasons explained above.

We can still continue discussion if you are still having problems (or have any comments/queries) after I close the issue now. I am happy to reopen it if we find any other relevant bugs.
